File: app/lib/features/stories/domain/life_story.dart
enum StoryStatus { pending, approved, rejected }

StoryStatus _statusFromJson(String raw) => switch (raw) {
      'approved' => StoryStatus.approved,
      'rejected' => StoryStatus.rejected,
      _ => StoryStatus.pending,
    };

/// One person's own account of their mental-health journey, submitted for
/// the public guide. Mirrors the backend's `LifeStory` — see
/// `mental_domain::life_story` for why there's no separate medication
/// field.
class LifeStory {
  final String id;
  final String body;
  /// ISO-639-1-ish code the backend detected at submission — compared
  /// against the viewer's own app language to decide whether the feed
  /// offers a translated copy. See `StoriesController.translationFor`.
  final String language;
  final String diagnosisSlug;
  final StoryStatus status;
  final bool crisisFlag;
  final DateTime createdAt;
  final bool anonymous;
  final int upvotes;
  final bool viewerUpvoted;

  /// Null for an anonymous story — the backend doesn't send an id to
  /// click through to, not just no name to render.
  final String? authorUserId;
  final String? authorDisplayName;
  final bool authorHasAvatar;

  const LifeStory({
    required this.id,
    required this.body,
    this.language = 'tr',
    required this.diagnosisSlug,
    required this.status,
    required this.crisisFlag,
    required this.createdAt,
    this.anonymous = true,
    this.upvotes = 0,
    this.viewerUpvoted = false,
    this.authorUserId,
    this.authorDisplayName,
    this.authorHasAvatar = false,
  });

  /// `/stories` (the feed) adds the vote tally and, when the author
  /// signed it, their identity. No status or crisis flag: every entry
  /// there is already approved.
  factory LifeStory.fromPublicJson(Map<String, dynamic> json) => LifeStory(
        id: json['id'] as String,
        body: json['body'] as String,
        language: json['language'] as String? ?? 'tr',
        diagnosisSlug: json['diagnosis_slug'] as String? ?? '',
        status: StoryStatus.approved,
        crisisFlag: false,
        createdAt: DateTime.parse(json['created_at'] as String),
        anonymous: json['anonymous'] as bool? ?? true,
        upvotes: json['upvotes'] as int? ?? 0,
        viewerUpvoted: json['viewer_upvoted'] as bool? ?? false,
        authorUserId: json['author_user_id'] as String?,
        authorDisplayName: json['author_display_name'] as String?,
        authorHasAvatar: json['author_has_avatar'] as bool? ?? false,
      );

  /// `/stories/mine` returns the full record, including status.
  factory LifeStory.fromJson(Map<String, dynamic> json) => LifeStory(
        id: json['id'] as String,
        body: json['body'] as String,
        diagnosisSlug: json['diagnosis_slug'] as String? ?? '',
        status: _statusFromJson(json['status'] as String),
        crisisFlag: json['crisis_flag'] as bool? ?? false,
        createdAt: DateTime.parse(json['created_at'] as String),
        anonymous: json['anonymous'] as bool? ?? true,
      );

  /// Local echo of a vote, so the row updates on tap instead of waiting
  /// for the feed to be refetched.
  LifeStory withVote({required bool upvoted}) => LifeStory(
        id: id,
        body: body,
        language: language,
        diagnosisSlug: diagnosisSlug,
        status: status,
        crisisFlag: crisisFlag,
        createdAt: createdAt,
        anonymous: anonymous,
        upvotes: upvotes + (upvoted ? 1 : -1),
        viewerUpvoted: upvoted,
        authorUserId: authorUserId,
        authorDisplayName: authorDisplayName,
        authorHasAvatar: authorHasAvatar,
      );
}

/// A submission as an admin sees it while moderating — identity included.
class AdminStoryView {
  final String id;
  final String userId;
  final String displayName;
  final String body;
  final StoryStatus status;
  final bool crisisFlag;
  final DateTime createdAt;

  const AdminStoryView({
    required this.id,
    required this.userId,
    required this.displayName,
    required this.body,
    required this.status,
    required this.crisisFlag,
    required this.createdAt,
  });

  /// Every account's `display_name` defaults to the same placeholder and
  /// is rarely changed, so it doesn't actually distinguish submitters in
  /// the moderation queue — a short, stable handle derived from the id
  /// does. Purely a display choice; nothing is sent or stored under it.
  String get handle => 'kullanıcı_${(userId.hashCode.abs() % 1000).toString().padLeft(3, '0')}';

  factory AdminStoryView.fromJson(Map<String, dynamic> json) => AdminStoryView(
        id: json['id'] as String,
        userId: json['user_id'] as String,
        displayName: json['display_name'] as String,
        body: json['body'] as String,
        status: _statusFromJson(json['status'] as String),
        crisisFlag: json['crisis_flag'] as bool? ?? false,
        createdAt: DateTime.parse(json['created_at'] as String),
      );
}

/// An already-published story a reader flagged for re-review.
class ReportedStory {
  final String reportId;
  final String? note;
  final DateTime reportedAt;
  final AdminStoryView story;

  const ReportedStory({
    required this.reportId,
    required this.note,
    required this.reportedAt,
    required this.story,
  });

  factory ReportedStory.fromJson(Map<String, dynamic> json) => ReportedStory(
        reportId: json['report_id'] as String,
        note: json['note'] as String?,
        reportedAt: DateTime.parse(json['reported_at'] as String),
        story: AdminStoryView.fromJson(json['story'] as Map<String, dynamic>),
      );
}

File: backend/crates/domain/src/life_story.rs
use chrono::{DateTime, Utc};
use serde::{Deserialize, Serialize};
use uuid::Uuid;

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum StoryStatus {
    Pending,
    Approved,
    Rejected,
}

impl StoryStatus {
    pub fn as_str(&self) -> &'static str {
        match self {
            StoryStatus::Pending => "pending",
            StoryStatus::Approved => "approved",
            StoryStatus::Rejected => "rejected",
        }
    }

    pub fn parse(raw: &str) -> Self {
        match raw {
            "approved" => StoryStatus::Approved,
            "rejected" => StoryStatus::Rejected,
            _ => StoryStatus::Pending,
        }
    }
}

/// A first-person account of someone's own mental-health journey —
/// submitted for the public guide rather than kept private like a
/// journal entry. Every submission starts `Pending`; only an admin's
/// approve/reject moves it, and the author can withdraw it at any status.
///
/// Deliberately free text with no structured "medication name" field: a
/// searchable directory of who recommends which drug reads as promoting
/// a specific product, which is what Turkey's regulation on advertising
/// medicinal products to the public restricts. A personal narrative that
/// happens to mention a medication in passing is a different thing.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct LifeStory {
    pub id: Uuid,
    pub user_id: Uuid,
    pub body: String,
    /// The catalog condition this story is about — required at
    /// submission, so the guide can filter the feed by diagnosis instead
    /// of it being one undifferentiated wall of text.
    pub diagnosis_slug: String,
    pub status: StoryStatus,
    /// Same keyword screen used for journal/chat, run at submission time
    /// so a crisis-flagged story gets extra scrutiny in the approval
    /// queue instead of blending in with the rest.
    pub crisis_flag: bool,
    /// When the author checked the "this will be shown to other users"
    /// box — a separate consent from account signup, since "visible to
    /// other users" is a materially different use of the data than a
    /// private journal entry.
    pub consented_at: DateTime<Utc>,
    /// Whether the feed hides who wrote this. Chosen per story rather
    /// than per account: the same person can want their name on one
    /// account of their life and not on another.
    #[serde(default = "default_anonymous")]
    pub anonymous: bool,
    /// ISO-639-1-ish code detected from the body at submission time (see
    /// [`detect_language`]). Drives the feed's auto-translate: a reader
    /// whose own account language differs from this gets a translated
    /// copy with a toggle back to the original — see
    /// `routes::stories::translate`.
    #[serde(default = "default_language")]
    pub language: String,
    pub reviewed_at: Option<DateTime<Utc>>,
    pub created_at: DateTime<Utc>,
}

fn default_anonymous() -> bool {
    true
}

fn default_language() -> String {
    "tr".to_string()
}

/// A deliberately small heuristic, not a statistical language-id model:
/// Turkish-specific letters (ç ğ ı ö ş ü, any case) are a strong enough
/// signal on their own, and their absence from text of any real length
/// is a strong enough signal the other way. Good enough to route
/// "should this be offered a translation", not meant as a general-purpose
/// classifier — a misdetected story just means one unnecessary (or one
/// skipped) translate button, never a wrong score or a lost submission.
/// Returns a free-form language code rather than an enum so adding a
/// third app language later is a matter of teaching this function one
/// more signal, not a schema change.
pub fn detect_language(text: &str) -> String {
    const TURKISH_CHARS: &[char] = &['ç', 'ğ', 'ı', 'ö', 'ş', 'ü', 'Ç', 'Ğ', 'İ', 'Ö', 'Ş', 'Ü'];
    if text.chars().any(|c| TURKISH_CHARS.contains(&c)) {
        return "tr".to_string();
    }

    // No Turkish-specific letters. Short text (a few words) is too little
    // signal either way, so default to Turkish — the app's primary
    // language — rather than guess. Longer plain-ASCII text with common
    // English function words is treated as English.
    let lower = text.to_lowercase();
    let word_count = lower.split_whitespace().count();
    let is_ascii_ish = text.chars().all(|c| c.is_ascii() || c.is_whitespace());
    const ENGLISH_MARKERS: &[&str] =
        &[" the ", " and ", " that ", " with ", " have ", " this ", " because ", " my "];
    let padded = format!(" {lower} ");
    let english_hits = ENGLISH_MARKERS.iter().filter(|m| padded.contains(*m)).count();

    if word_count >= 6 && is_ascii_ish && english_hits >= 1 {
        "en".to_string()
    } else {
        "tr".to_string()
    }
}

/// A feed row: the story plus everything the reader's copy of it needs —
/// vote tally, whether *they* voted, and the author's identity when the
/// story isn't anonymous. Assembled in one query rather than a lookup
/// per story.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct StoryFeedItem {
    pub story: LifeStory,
    pub upvotes: u32,
    pub viewer_upvoted: bool,
    pub author_display_name: String,
    pub author_has_avatar: bool,
}

/// A reader flagging an already-approved story back for re-review.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct LifeStoryReport {
    pub id: Uuid,
    pub story_id: Uuid,
    pub reporter_user_id: Uuid,
    pub note: Option<String>,
    pub created_at: DateTime<Utc>,
}
